services/worker/src/worker/job_manager.py

```diff
@@ -0,0 +1,308 @@
+# SPDX-License-Identifier: Apache-2.0
+# Copyright 2022 The HuggingFace Authors.
+
+import logging
+from http import HTTPStatus
+from typing import Literal, Optional
+
+from libcommon.config import CommonConfig
+from libcommon.dataset import DatasetNotFoundError, get_dataset_git_revision
+from libcommon.exceptions import CustomError
+from libcommon.processing_graph import ProcessingGraph, ProcessingStep
+from libcommon.simple_cache import (
+    DoesNotExist,
+    get_response_without_content_params,
+    upsert_response_params,
+)
+from libcommon.state import DatasetState
+from libcommon.utils import JobInfo, JobParams, Priority, Status, orjson_dumps
+
+from worker.common_exceptions import (
+    JobManagerCrashedError,
+    JobManagerExceededMaximumDurationError,
+    NoGitRevisionError,
+    ResponseAlreadyComputedError,
+    TooBigContentError,
+    UnexpectedError,
+)
+from worker.config import AppConfig, WorkerConfig
+from worker.job_runner import JobRunner
+
+# List of error codes that should trigger a retry.
+ERROR_CODES_TO_RETRY: list[str] = ["ClientConnectionError"]
+
+
+class JobManager:
+    """
+    A job manager is a class that handles a job runner compute, for a specific processing step.
+
+    Args:
+        job_info (:obj:`JobInfo`):
+            The job to process. It contains the job_id, the job type, the dataset, the config, the split
+            the force flag, and the priority level.
+        common_config (:obj:`CommonConfig`):
+            The common config.
+        processing_step (:obj:`ProcessingStep`):
+            The processing step to process.
+    """
+
+    job_id: str
+    dataset: str
+    job_params: JobParams
+    force: bool
+    priority: Priority
+    worker_config: WorkerConfig
+    common_config: CommonConfig
+    processing_step: ProcessingStep
+    _dataset_git_revision: Optional[str] = None
+    job_runner: JobRunner
+
+    def __init__(
+        self,
+        job_info: JobInfo,
+        app_config: AppConfig,
+        job_runner: JobRunner,
+        processing_graph: ProcessingGraph,
+    ) -> None:
+        self.job_info = job_info
+        self.job_type = job_info["type"]
+        self.job_id = job_info["job_id"]
+        self.force = job_info["force"]
+        self.priority = job_info["priority"]
+        self.job_params = job_info["params"]
+        self.dataset = self.job_params["dataset"]
+        self.common_config = app_config.common
+        self.worker_config = app_config.worker
+        self.job_runner = job_runner
+        self.processing_graph = processing_graph
```

Comment:
could we add it to the declared fields above? (and the other missing ones, if there are some?)

Reply:
Thanks, I added it on https://github.com/huggingface/datasets-server/pull/1146/commits/9b6bf8e0f8b93176cd3f8dd7e36025d62edb3305